src/core/experimental/array.d

```diff
@@ -0,0 +1,2299 @@
+        bool _expand(bool isShared, ref void[] b, size_t delta)
+        {
+            return isShared ?  sharedAllocator.expand(b, delta) : localAllocator.expand(b, delta);
+        }
+    }
+
+    @nogc nothrow pure
+    void pureDispose(T)(bool isShared, T[] b)
+    {
+        return (cast(void function(bool, T[]) @nogc nothrow pure)(&_dispose!(T)))(isShared, b);
+    }
+
+    @nogc nothrow
+    void _dispose(T)(bool isShared, T[] b)
+    {
+        return isShared ?  sharedAllocator.dispose(b) : localAllocator.dispose(b);
+    }
+}
+
+///
+struct Array(T)
+{
+private:
+
+    import core.atomic : atomicOp;
+
+    T[] payload;
+    Unqual!T[] support;
+
+    version(unittest)
+    {
+    }
+    else
+    {
+        alias localAllocator = shared PrefixAllocator.instance;
+        alias sharedAllocator = shared PrefixAllocator.instance;
+    }
+
+    static enum double capacityFactor = 3.0 / 2;
+    static enum initCapacity = 3;
+    bool isShared;
+
+    @trusted
+    auto pref() const
+    {
+        assert(support !is null);
+        if (isShared)
+        {
+            return sharedAllocator.prefix(support);
+        }
+        else
+        {
+            return localAllocator.prefix(support);
+        }
+    }
+
+    private size_t _opPrefix(string op, T)(const T[] _support, size_t val) const
+    {
+        assert(_support !is null);
+        if (isShared)
+        {
+            return cast(size_t)(atomicOp!op(*cast(shared size_t *)&sharedAllocator.prefix(_support), val));
+        }
+        else
+        {
+            mixin("return cast(size_t)(*cast(size_t *)&localAllocator.prefix(_support)" ~ op ~ "val);");
+        }
+    }
+
+    @nogc nothrow pure @trusted
+    size_t opPrefix(string op, T)(const T[] _support, size_t val) const
+    if ((op == "+=") || (op == "-="))
+    {
+
+        return (cast(size_t delegate(const T[], size_t) const @nogc nothrow pure)(&_opPrefix!(op, T)))(_support, val);
+    }
+
+    @nogc nothrow pure @trusted
+    size_t opCmpPrefix(string op, T)(const T[] _support, size_t val) const
+    if ((op == "==") || (op == "<=") || (op == "<") || (op == ">=") || (op == ">"))
+    {
+        return (cast(size_t delegate(const T[], size_t) const @nogc nothrow pure)(&_opPrefix!(op, T)))(_support, val);
+    }
+
+    @nogc nothrow pure @trusted
+    void addRef(SupportQual, this Q)(SupportQual _support)
+    {
+        assert(_support !is null);
+        cast(void) opPrefix!("+=")(_support, 1);
+    }
+
+    void delRef(Unqual!T[] _support)
+    {
+        // Will be optimized away, but the type system infers T's safety
+        if (0) { T t = T.init; }
+
+        assert(_support !is null);
+        if (opPrefix!("-=")(_support, 1) == 0)
+        {
+            () @trusted {
+                version(unittest)
+                {
+                    pureDispose(isShared, _support);
+                }
+                else
+                {
+                    localAllocator.dispose(_support);
+                }
```

Comment:
What's the motivation for this hack?

Reply:
The `version(unittest)` is needed in order to use an allocator that can collect information about the number of bytes used, during testing. I know this is not the ideal approach, but I couldn't figure something better.
We need the unittest version of the allocator, as this has caught a couple of implementation bugs in the past. I would strongly advise against removing it.